### PR TITLE
Enable custom vocabularies via ConfigMaps

### LINKED ISSUE #135: Allow passing vocabulary files to the pods
We have several instances that share the same docker image. It'd be nice to configure the vocabularies each one loads using the helm chart.

Maybe something like this:
```
invenio:
  ...
  vocabularies:
    resource_types.yaml: | 
        - id: publication
          icon: file alternate
          props:
            csl: report
            datacite_general: Text
            datacite_type: ""   
            openaire_resourceType: "0017"
            openaire_type: publication
            eurepo: info:eu-repo/semantics/other
            schema.org: https://schema.org/CreativeWork
            subtype: ""
            type: publication
            marc21_type: publication
            marc21_subtype: ""
          title:
            en: Publication
          tags:
            - depositable
            - linkable
          ...
```

### PR DESCRIPTION
**Pull request by egabancho**
* Adds the option to customize vocabularies (closes #135)
* Adds proper labels everywhere

### CODE REVIEW

**Comment by lindhe** [comment]
Can vocabularies really fit in a ConfigMap? I was under the impression that they are often pretty big.

File: charts/invenio/templates/logstash/filebeat-configmap.yaml [hunk 1]
```diff
@@ -4,6 +4,8 @@ apiVersion: v1
 kind: ConfigMap
 metadata:
   name: filebeat-config
+  labels:
```
Comment (lindhe) [inline charts/invenio/templates/logstash/filebeat-configmap.yaml:7]: I like that you try to fix the label situation, but I think this belongs in another PR. It's hard to review multiple things in one PR on GitHub.

Comment (lindhe) [reply charts/invenio/templates/logstash/filebeat-configmap.yaml:7]: But again: I _really_ appreciate how good you are at creating separate commits for each change. It's so nice to see someone that creates atomic commits for a change! ❤️

Comment (egabancho) [reply charts/invenio/templates/logstash/filebeat-configmap.yaml:7]: I know; it's just faster for me to work like that sometimes. I'm making some of the changes because I feel they need to be done, not because we actually need them.

On the `Files changed` tab, you can select the commit you want to check. That may help.

Comment (egabancho) [reply charts/invenio/templates/logstash/filebeat-configmap.yaml:7]: Also, I just realized that I based this one on top of the other PR ... that's probably also confusing 😂

Comment (lindhe) [reply charts/invenio/templates/logstash/filebeat-configmap.yaml:7]: Yeah, the changes that are from other PRs must for sure be removed here, very hard to review otherwise. 🙂

Comment (egabancho) [reply charts/invenio/templates/logstash/filebeat-configmap.yaml]: Done, I moved the label stuff [here](https://github.com/inveniosoftware/helm-invenio/pull/138) and rebase.

File: charts/invenio/templates/install-init-job.yaml [hunk 2]
```diff
@@ -56,10 +56,24 @@ spec:
             secretKeyRef:
               name: invenio-secrets
               key: INVENIO_CSRF_SECRET_SALT
+        volumeMounts:
+          {{- if .Values.invenio.vocabularies }}
```
Comment (lindhe) [inline charts/invenio/templates/install-init-job.yaml:60]: I think this if is redundant or misplaced. I have not run it myself to verify, but I think `range` will run 0 times (not error) if `.Values.invenio.vocabularies` is null or has no elements.

On the other hand, if `.Values.invenio.vocabularies` is null then you will still have the `volumeMounts` field printed with empty contents (since the `volumeMounts` is defined outside of the `if`).

File: charts/invenio/templates/install-init-job.yaml [hunk 3]
```diff
@@ -56,10 +56,24 @@ spec:
...
+            mountPath: "/opt/invenio/var/instance/app_data/vocabularies/{{ $key }}"
+            subPath: "{{ $key }}"
+            {{- end }}
+          {{- end }}
         {{- if .Values.web.resources }}
         resources: {{- toYaml .Values.web.resources | nindent 10 }}
         {{- end }}
       restartPolicy: OnFailure
+      volumes:
+        {{- if .Values.invenio.vocabularies}}
```
Comment (lindhe) [inline charts/invenio/templates/install-init-job.yaml:72]: If `.Values.invenio.vocabularies` is null, `volumes` will be empty. Better to not include `volumes` if not used?

File: charts/invenio/templates/vocabularies-configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+kind: ConfigMap
+metadata:
+  name: invenio-vocabularies
+  labels:
+    {{- include "invenio.labels" . | nindent 4 }}
+data:
+{{- range $key, $value := .Values.invenio.vocabularies }}
+  {{ $key }}: |
+  {{- $value | nindent 4 }}
+{{- end }}
```
Comment (lindhe) [inline charts/invenio/templates/vocabularies-configmap.yaml:12]: Not a bug, I just want to highlight that this logic can probably be replaced by `{{- .Values.invenio.vocabularies | toYaml | nindent 2 }}` or something like that. No need to iterate over the objects.

File: charts/invenio/templates/web-deployment.yaml [hunk 5]
```diff
@@ -126,6 +126,13 @@ spec:
           - name: kerberos-credentials-cache
             mountPath: /tmp
           {{- end }}
+          {{- if .Values.invenio.vocabularies }}
```
Comment (lindhe) [inline charts/invenio/templates/web-deployment.yaml:129]: I think this `if` is redundant.

File: charts/invenio/templates/worker-beat-deployment.yaml [hunk 6]
```diff
@@ -33,6 +33,13 @@ spec:
         - mountPath: /opt/invenio/var/instance/data
           name: shared-volume
         {{- end }}
+        {{- if .Values.invenio.vocabularies }}
```
Comment (lindhe) [inline charts/invenio/templates/worker-beat-deployment.yaml:36]: I think this `if` is redundant.

File: charts/invenio/templates/worker-deployment.yaml [hunk 7]
```diff
@@ -109,6 +109,13 @@ spec:
             - name: kerberos-credentials-cache
               mountPath: /tmp
             {{- end }}
+            {{- if .Values.invenio.vocabularies }}
```
Comment (lindhe) [inline charts/invenio/templates/worker-deployment.yaml:112]: I think this `if` is redundant.

File: charts/invenio/templates/vocabularies-configmap.yaml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{{- if .Values.invenio.vocabularies}}
```
Comment (lindhe) [inline charts/invenio/templates/vocabularies-configmap.yaml:1]:
```suggestion
{{- if .Values.invenio.vocabularies }}
```

File: charts/invenio/templates/vocabularies-configmap.yaml [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: invenio-vocabularies
+data:
+{{- .Values.invenio.vocabularies | toYaml | nindent 2 }}
+{{- end }}
```
Comment (lindhe) [inline charts/invenio/templates/vocabularies-configmap.yaml:8]:
```suggestion
{{- with .Values.invenio.vocabularies }}
apiVersion: v1
kind: ConfigMap
metadata:
  name: invenio-vocabularies
data:
  {{- toYaml . | nindent 2 }}
{{- end }}
```

Comment (lindhe) [reply charts/invenio/templates/vocabularies-configmap.yaml:8]: Sorry for being nit-picky. But I like reducing code repetition. It's so easy to make typos in the future if multiple places needs to be changed.

File: charts/invenio/templates/web-deployment.yaml [hunk 10]
```diff
@@ -274,6 +279,11 @@ spec:
       - name: kerberos-credentials-cache
         emptyDir: {}
       {{- end }}
+      {{- if .Values.invenio.vocabularies}}
```
Comment (lindhe) [inline charts/invenio/templates/web-deployment.yaml:282]:
```suggestion
      {{- if .Values.invenio.vocabularies }}
```

File: charts/invenio/templates/worker-beat-deployment.yaml [hunk 11]
```diff
@@ -122,6 +127,11 @@ spec:
         persistentVolumeClaim:
           claimName: {{ .Values.persistence.name }}
       {{- end }}
+      {{- if .Values.invenio.vocabularies}}
```
Comment (lindhe) [inline charts/invenio/templates/worker-beat-deployment.yaml:130]:
```suggestion
      {{- if .Values.invenio.vocabularies }}
```

File: charts/invenio/templates/worker-deployment.yaml [hunk 12]
```diff
@@ -160,7 +165,11 @@ spec:
         - name: kerberos-credentials-cache
           emptyDir: {}
         {{- end }}
-
+        {{- if .Values.invenio.vocabularies}}
```
Comment (lindhe) [inline charts/invenio/templates/worker-deployment.yaml:168]:
```suggestion
        {{- if .Values.invenio.vocabularies }}
```

File: charts/invenio/templates/install-init-job.yaml [hunk 13]
```diff
@@ -56,10 +56,22 @@ spec:
...
+          {{- range $key, $value := .Values.invenio.vocabularies }}
+          - name: vocabularies
+            mountPath: "/opt/invenio/var/instance/app_data/vocabularies/{{ $key }}"
+            subPath: "{{ $key }}"
+          {{- end }}
         {{- if .Values.web.resources }}
         resources: {{- toYaml .Values.web.resources | nindent 10 }}
         {{- end }}
       restartPolicy: OnFailure
+      {{- if .Values.invenio.vocabularies}}
```
Comment (lindhe) [inline charts/invenio/templates/install-init-job.yaml:69]:
```suggestion
      {{- if .Values.invenio.vocabularies }}
```

File: charts/invenio/templates/install-init-job.yaml [hunk 14]
```diff
@@ -56,10 +56,22 @@ spec:
             secretKeyRef:
               name: invenio-secrets
               key: INVENIO_CSRF_SECRET_SALT
+        volumeMounts:
+          {{- range $key, $value := .Values.invenio.vocabularies }}
+          - name: vocabularies
+            mountPath: "/opt/invenio/var/instance/app_data/vocabularies/{{ $key }}"
+            subPath: "{{ $key }}"
```
Comment (lindhe) [inline charts/invenio/templates/install-init-job.yaml:63]: Is `/opt/invenio/var/instance/app_data/vocabularies/` generally prepopulated by the image, so other files exists there without mount?

If so, we do need to use `subPath` like you do here, so the other files are not overwritten by the volume mount.

On the other hand, if that path is not prepopulated by the image, then I it would be advantageous to drop `subPath`. Mostly because `subPath` mounts are not updated without a pod restart, but also because that would let us drop the entire `range` since each field in a `ConfigMap` gets mounted then.

(same comment for the similar places in this PR).

**Review by lindhe — CHANGES_REQUESTED** [review]
Sorry for being so slow to review. I hate being a blocker.

Just have some minor comments here, mostly style changes. As soon as they are fixed (or answered) I approve this PR. 👍

File: charts/invenio/templates/vocabularies-configmap.yaml (hunk 9 above)
Comment (egabancho) [reply charts/invenio/templates/vocabularies-configmap.yaml]: You **have to be** nit-picky, that's what we pay you for 😜

File: charts/invenio/templates/install-init-job.yaml (hunk 14 above)
Comment (egabancho) [reply charts/invenio/templates/install-init-job.yaml:64]: In general, the folder already existing inside the default docker image, https://github.com/inveniosoftware/cookiecutter-invenio-rdm/blob/master/%7B%7Bcookiecutter.project_shortname%7D%7D/Dockerfile#L20 and https://github.com/inveniosoftware/cookiecutter-invenio-rdm/tree/master/%7B%7Bcookiecutter.project_shortname%7D%7D/app_data/vocabularies.
I say _in general_ because one can always change the docker image and not add it ...

Comment (egabancho) [reply charts/invenio/templates/install-init-job.yaml:64]: It is more or less hardcoded here https://github.com/inveniosoftware/invenio-rdm-records/blob/master/invenio_rdm_records/fixtures/__init__.py#L23

File: charts/invenio/templates/vocabularies-configmap.yaml [hunk 15]
```diff
@@ -0,0 +1,10 @@
+{{- with .Values.invenio.vocabularies }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: invenio-vocabularies
+  labels:
+    {{- include "invenio.labels" . | nindent 4 }}
```
Comment (lindhe) [inline charts/invenio/templates/vocabularies-configmap.yaml:7]: Please verify by running `helm template` if `include "invenio.labels" .` works or if we need to use `include "invenio.labels" $` instead (since `with` changes the scope). I'm not sure, we need to check.

File: charts/invenio/templates/install-init-job.yaml (hunk 14 above)
Comment (lindhe) [reply charts/invenio/templates/install-init-job.yaml:64]: Thanks for checking. Then I think we must use `subPath`. Thus, this part looks good in its current form.

File: charts/invenio/templates/vocabularies-configmap.yaml (hunk 15 above)
Comment (egabancho) [reply charts/invenio/templates/vocabularies-configmap.yaml]: Good catch!

**Review by lindhe — APPROVED** [review]
